op-conductor/rpc/execution_proxy.go

```diff
@@ -0,0 +1,36 @@
+package rpc
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var ExecutionRPCNamespace = "eth"
+
+// ExecutionProxyBackend implements an execution rpc proxy with a leadership check before each call.
+type ExecutionProxyBackend struct {
+	log    log.Logger
+	con    conductor
+	client *ethclient.Client
+}
+
+var _ ExecutionProxyAPI = (*ExecutionProxyBackend)(nil)
+
+func NewExecutionProxyBackend(log log.Logger, con conductor, client *ethclient.Client) *ExecutionProxyBackend {
+	return &ExecutionProxyBackend{
+		log:    log,
+		con:    con,
+		client: client,
+	}
+}
+
+func (api *ExecutionProxyBackend) BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error) {
+	if !api.con.Leader(ctx) {
+		return nil, ErrNotLeader
+	}
+	return api.client.BlockByNumber(ctx, number)
```

Comment:
If we check leader and then query the block, then one failure scenario could be that leadership changed to another node after the check, and current sequencer is producing wrong blocks that shouldn't be retrieved into batcher.

Move leadership check after retrieving block number makes sure that the block information we get is from a leader which implicitly should have the correct block.

```suggestion
    block, err := api.client.BlockByNumber(ctx, number)
    if err != nil {
        return nil, err
    }
    
	if !api.con.Leader(ctx) {
		return nil, ErrNotLeader
	}
	
	return block, nil
```

Reply:
Good catch!